### ISSUE TITLE
Object is possibly undefined` error in TypeScript 4.1.2

### ISSUE THREAD

**Issue by doberkofler** [microsoft/TypeScript#41612]
**TypeScript Version:**  4.1.2

**Search Terms:** Object is possibly undefined

**Code**

```ts
function main(values: Array<string>) {
    for (let i = 0; i < values.length; i++) {
        if (typeof values[i] === 'string' && values[i].length > 0) {
            console.log(i);
        }
    }
}
```

**Expected behavior:**
No error

**Actual behavior:**
TypeScript 4.1.2 reports a `Object is possibly undefined` error in `values[i].length` but given the first condition the object must be defined as it is of type 'string'

**Playground Link:** https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=4.1.0-beta#code/GYVwdgxgLglg9mABAWwIYzACgG6oDYgCmAzgFyICCATlagJ4A8xUVGA5gHwCUiA3gFCIhiYHCqJMeQlEQxEAXkQAGANyzEDRLgIkAdFLBsoACzUwA1OZ4Dht2cAlQ6AB0JwH2osQDaMALoK8ooA5MyshsGIAGRRWvhevn76hIYmiBzK1oJ2ORAIxHBS+nBsmDBcKtk5AL5ViLW1QA

**Related Issues:** no

**Comment by MartinJohns** [comment]
Duplicate of #10530. Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

**Comment by doberkofler** [comment]
@MartinJohns Sorry for the duplicate, but I would not have found it and really looked. I just wanted to add that this specific use case of the problem is the main cause for an immense number of "false errors" when activating the new `noUncheckedIndexedAccess` flag in my code base.

**Comment by RyanCavanaugh** [comment]
This was one of the caveats that kept the flag from being implemented for so long. We don't have any way to ascertain whether `i` or the indexed array has changed between any two arbitrary expressions, so the only "safe" thing to do is always include `undefined`. You can write
```ts
function main(values: Array<string>) {
    for (let i = 0; i < values.length; i++) {
        const el = values[i]'
        if (typeof el === 'string' && el.length > 0) {
            console.log(i);
        }
    }
}
```

**Comment by thecotne** [comment]
@RyanCavanaugh this would work as well

```ts
function main(values: ReadonlyArray<string>) {
    for (let i = 0; i < values.length; i++) {
        if (typeof values[i] === 'string' && values[i].length > 0) {
            console.log(i);
        }
    }
}
```

since `ReadonlyArray` can't magically change what indexes are in the set and what are not

**Comment by doberkofler** [comment]
@thecotne Just to make sure: it should but it does not. Correct?

**Comment by thecotne** [comment]
@doberkofler sorry for confusion. it should but it does not.

https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=4.1.0-beta&ssl=1&ssc=31&pln=1&pc=23#code/GYVwdgxgLglg9mABAWwIYzACgG6oDYgCmAzgFyIBKhqAJgngJ4CCATi6gwDzFQsYDmAPgCUiAN4AoRNMTA4LRJjyEoiGIgC8iAAwBuNYk6JcBEgDplYflAAW+mAGoHoyTLdrgiqAwAOhOJ4mRMQA2jAAupoaWgDkPHxWMYgAZMnG+MFh4RaEVraIgjouUu6lEAjEcMoWcPyYMMK6JaUAvs2IbS1AA

**Comment by Skillz4Killz** [comment]
Have you considered using a for of or forEach loop?
```ts
function main(values: Array<string>) {
    for (const value of values){
        if (value) {
            console.log(value);
        }
    }
}

function main(values: Array<string>) {
    values.forEach(value => {
         if (value) {
             console.log(value);
         }
    })
}
```

**Comment by doberkofler** [comment]
@Skillz4Killz You are absolutely correct. The code snipped is just an example for the generic problem where the array element is first tested and then used if available.

**Comment by saperdadsk** [comment]
@doberkofler Consider the following (contrived) example though
```
function main(values: Array<string>) {
    for (let i = 0; i < values.length; i++) {
        const incrementI = () => {i = i + 1;}
        if (typeof values[i] === 'string' && incrementI() && values[i].length > 0) {
            console.log(i);
        }
    }
}
```
values[i].length is no longer guaranteed to be defined, and TS has no good way to detecting that change.

**Comment by RyanCavanaugh** [comment]
> since ReadonlyArray can't magically change what indexes are in the set and what are not

This isn't true. A `ReadonlyArray` might be an alias for an underlying mutable array.

**Comment by peterholak** [comment]
> This was one of the caveats that kept the flag from being implemented for so long. We don't have any way to ascertain whether `i` or the indexed array has changed between any two arbitrary expressions, so the only "safe" thing to do is always include `undefined`.

But doesn't narrowing on all object properties work in the "unsafe way" in TypeScript anyway? E.g.

```typescript
type X = { word: string|undefined }

function bad(obj: X) {
    if (obj.word !== undefined) {
        mutate(obj)
        return obj.word.length
    }
    return undefined
}

function mutate(obj: X) {
    obj.word = undefined
}

console.log(bad({ word: 'hello' }))
```

Produces a run-time error, but no TypeScript errors.

**Comment by RyanCavanaugh** [comment]
The entire point of the flag was to be more safe than TS's normal behavior, so it being more restrictive than the default is a not undesirable aspect.

**Comment by yume-chan** [comment]
I know TS doesn't recognize this pattern, even for normal optional properties:

```ts
interface Foo {
  bar: string | undefined;
}

declare const foo: Foo;

if (foo.bar) {
  foo.bar.toLowerCase(); // OK
}

const bar = 'bar';
if (foo[bar]) {
  foo[bar].toLowerCase(); // Object is possibly 'undefined'.
}
```
([Playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgGIHt3IN4ChnIBGcUAXMgM5hSgDmyAPsgK4gAmEMoEbA3LgF9cuDggA2JFAnQgqyGJnIZ0-XMBjIAFAvQA6YlACUOfPMz6SusOgAy6AO7QAwnAoRNh3sgD035AHkAaUFhaVkwIhJkAF5kAHIDOP51LR0AbQMAXWM8AnSsq1sHZ1d3Tx8-f0IAKwgECOAKZAAHdAoKYEIxAE941g4uEB443UEgA))

However I can't understand why:

If it's because a variable (`bar` in this case) may change between two accesses, since `bar` is a constant, it's not possible.

If you say the object (`foo` in this case) may change, then why does accessing with literals work? I expect they (access with literals and variables) both work or not work, and I prefer both not.

**Comment by doberkofler** [comment]
@RyanCavanaugh 
> The entire point of the flag was to be more safe than TS's normal behavior, so it being more restrictive than the default is a not undesirable aspect.

I understand that this option is restrictive and the only reason for my SR was to better understand how to take advantage of it. At least in my "real world" scenario, it seems almost impossible using it, as it leads to reject thousands of code lines that are actually valid. I'm still wondering, if maybe just be identifying some common use cases that are legal, the usability of this interesting new option might be dramatically improved.

**Comment by radekmie** [comment]
@yume-chan It's not enough for `bar` to be `const`. Consider this:
```ts
const bar = { toString: Math.random };
({ [bar]: 1 }) // { '0.9271219051493171': 1 }
({ [bar]: 1 }) // { '0.07489693725883417': 1 }
```

It may work for `string`, `number`, `symbol`, and maybe a few others.

**Comment by peterholak** [comment]
> The entire point of the flag was to be more safe than TS's normal behavior, so it being more restrictive than the default is a not undesirable aspect.

I appreciate having that option, but it mixes together two different concerns in an inseparable way

1. safety in the face of mutating data
2. safety for indexed array access (even on immutable data)

I think there is a good chunk of people that would like having option 2. without also being forced into option 1 (as it is much more rare, while 2. is very common).

Even so, it is a valuable improvement, just not as useful as it could be.

**Comment by MartinJohns** [comment]
@peterholak Option two would require a way to represent immutable types in TypeScript. This is currently not possible. It's something I personally really want, but it's nothing in the foreseeable future. Issues to follow would be #14909, #16317, #17181.

**Comment by peterholak** [comment]
@MartinJohns By option 2 I just meant assuming that the array and index don't change within a block (maybe unless there is an explicit assignment within the same block), and doing the narrowing anyway, just like it works for non-indexed access (https://github.com/microsoft/TypeScript/issues/41612#issuecomment-731515670).

It would be unsafe if the data changes in other ways, but that's a good trade-off for many cases (especially compared to not adding `|undefined` to indexed access at all like in previous versions).

**Comment by yume-chan** [comment]
@radekmie 

1. Your code is valid JavaScript, but TS already disallow it.
2. I have already said "`bar` in this case" so I just want to focus on this case
3. I think your comment is more like "TS shouldn't do anything that's not 100% safe", however in my second theory TS has already accepted something "unsafe", so the point is "what's the possiblility for developers (at all skill levels) to make such a mistake" and "is it suitable to treat all such scenario as an error"

**Comment by radekmie** [comment]
@yume-chan I just wanted to emphasize that being `const`-constant is not enough, that's all. I agree with 3 though.

**Comment by karlhorky** [comment]
@RyanCavanaugh why was this closed? Was there something that resolve this?

**Comment by peterholak** [comment]
For what it's worth, I still think this is 2 separate features mixed under the same flag. One is *"stricter narrowing"* (but not everywhere) and the other is *"adding |undefined to indexed access"*.
